File: 230        Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        def pushFullleft(root, stack):
            while root:
                stack.append(root)
                root = root.left
                
        stack = []
        pushFullleft(root, stack)
        
        node = None
        while stack and k > 0:
            node = stack.pop()
            k -= 1
            if node.right:
                pushFullleft(node.right, stack)
                
        return node.val

File: 230. Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void pushFullLeft(TreeNode* root, vector<TreeNode*>& stack) {
        while (root) {
            stack.push_back(root);
            root = root->left;
        }
    }
    int kthSmallest(TreeNode* root, int k) {
        vector<TreeNode*> stack;
        pushFullLeft(root, stack);
        while (stack.size() > 0) {
            auto node = stack.back();
            stack.pop_back();
            k -= 1;
            if (k == 0) {
                return node->val;
            }
            if (node->right) {
                pushFullLeft(node->right, stack);
            }
        }
    }
};
